### PR TITLE
added check for if the pubmed_url updated or not

### PR DESCRIPTION
**Pull request by HazalCiplak**
https://github.com/elifesciences/data-hub-issues/issues/262

### CODE REVIEW

File: notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb [hunk 1]
```diff
@@ -77,102 +75,124 @@
...
-   "metadata": {},
-   "outputs": [],
    "source": [
     "remaining_editor_parsed_pubmed_links_df = editor_parsed_pubmed_links_df[\n",
     "    ~editor_parsed_pubmed_links_df['person_id'].isin(existing_editor_ids_set)\n",
     "]\n",
+    "\n",
+    "remaining_editor_parsed_pubmed_links_df = remaining_editor_parsed_pubmed_links_df.append(\n",
+    "    editor_parsed_pubmed_links_df[\n",
+    "        ~editor_parsed_pubmed_links_df['pubmed_url'].isin(existing_editor_pubmed_url_set)\n",
```
Comment (de-code) [inline notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb:171]: In practise this will probably work most of the times, but it isn't 100% correct I believe.
For example, what if the editorial team notices that they had the URLs the wrong way around for two editors - then the URL would be in the set but not for the correct editor.

Maybe it will be worth creating a function a PeerScout related Python package in order to add tests?

For this particular case, one could have a set with tuple of `person_id` and `pubmed_id` and then check whether the set contains that. But that wouldn't be too helpful when adding imported timestamp.

The solution that would come to mind, is to use `merge` to left-join the two DataFrames. Then you can do simple Pandas operations, e.g. something like that:

```python
merged_editor_parsed_pubmed_links_df = editor_parsed_pubmed_links_df.merge(
    existing_editor_ids_and_pubmed_url_df,
    how='left',
    on='person_id',
    suffixes=('', '_existing')
)

remaining_editor_parsed_pubmed_links_df = merged_editor_parsed_pubmed_links_df[
    (merged_editor_parsed_pubmed_links_df['pubmed_url'] != merged_editor_parsed_pubmed_links_df['pubmed_url_existing'])
    | (merged_editor_parsed_pubmed_links_df['imported_timestamp'] < ...)
].drop(columns=[...])
```

Comment (HazalCiplak) [reply notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb:171]: If the URLs are in the wrong way around then the editorial team will realise and correct them, after that when the job run again it will be corrected in the dataframe too. In current solution it will cause duplication but because of we are planing to create the view after that step I was planing to keep just latest record for one editor. If I understand correctly it already will solve the problem? Between merging or appending will not be any difference in the case we use view step?

Comment (de-code) [reply notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb:171]: Apologies, I may not have explained it well. The merging is not related to merging in the output table. It will be appended either way. Here the merging is just a useful tool to compare between the new url and url from the output table.

Checking the existence of the person id in the set is fine. But checking the URL doesn't seem correct, because it is important whether the URL for that particular editor with the same person id has changed.

Perhaps it is easier to explain in person?

Comment (HazalCiplak) [reply notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb:171]: > it is important whether the URL for that particular editor with the same person id has changed.

Ah I see now :) OK thank you! I will look at it in this point of view.

File: notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb [hunk 2]
```diff
@@ -77,102 +75,124 @@
...
    "source": [
     "_sql = (\n",
-    "    'SELECT person_id FROM `{table}`'\n",
-    ").format(table=editor_parsed_pubmed_ids_table_name)\n",
+    "    '''\n",
+    "    SELECT\n",
+    "        person_id,\n",
+    "        pubmed_url,\n",
+    "        provenance.imported_timestamp as imported_timestamp\n",
+    "    FROM `{table}`\n",
```
Comment (de-code) [inline notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb:110]: Another thought that came to mind is that the table here will probably need to be the view that you were going to discuss. (Because we want to compare it to the latest version)

File: notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb [hunk 3]
```diff
@@ -77,102 +75,179 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "source": [
+    "editors_with_not_currently_updated_info = pd.DataFrame()\n",
+    "temp_df = pd.DataFrame()\n",
+    "for index, row in merged_editor_parsed_pubmed_links_df.iterrows():\n",
+    "    day_diff = abs(\n",
+    "        datetime.date(row['imported_timestamp'])-datetime.date(datetime.today())\n",
```
Comment (de-code) [inline notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb:203]: can't you do something like?

```python
editors_with_not_currently_updated_info_df = merged_editor_parsed_pubmed_links_df[
    (pd.Timestamp.now() - merged_editor_parsed_pubmed_links_df['imported_timestamp']).dt.days > 15
]
```

File: notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb (hunk 2 above)
Comment (de-code) [reply notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb]: Just mentioning that before merging this PR, we will need to make the following steps work with the changed output table.
If we are using a (sort-of) "materialized view" for that, then you could use that here too.

File: notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb (hunk 3 above)
Comment (de-code) [reply notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb:203]: Minor: "recently updated" may be slightly better

File: notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb [hunk 4]
```diff
@@ -77,102 +75,188 @@
...
+   "outputs": [],
+   "metadata": {}
   },
   {
    "cell_type": "code",
    "execution_count": null,
-   "metadata": {},
+   "source": [
+    "remaining_editor_parsed_pubmed_links_df = pd.concat([\n",
+    "    new_added_editors_df,\n",
```
Comment (de-code) [inline notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb:234]: One disadvantage of concatenating dataframes from different conditions is that you would want to make sure that there is no overlap. Otherwise the same person id could be included multiple times. That is probably the case for the first two, but not sure about `editors_with_not_currently_updated_info`? (e.g. if it had changed now, but also hasn't been updated for a while)

If you instead just select the rows, either using a single more complex condition, or chained, then this might be easier.

Perhaps an alternative, if you don't like the dataframe selection, could also be to have a final selection based on a set of person ids, similar to before. Something like:

```python
selected_person_ids = (
    set(new_added_editors_df['person_id'))
    | set(editors_with_changed_pubmed_url_df['person_id'))
    | set(editors_with_not_currently_updated_info['person_id'))
)
remaining_editor_parsed_pubmed_links_df = editor_parsed_pubmed_links_df[
    editor_parsed_pubmed_links_df['person_id'].isin(selected_person_ids)
]
```

**Comment by de-code** [comment]
I know that the notebook was lacking unit tests before. And my main excuse is that it was partially copied from previous "research" notebooks and that it was still a prototype. Sorry :-(

But given that there are some edge cases that are difficult to spot in a code review, it might really be useful to have unit tests. e.g. there could be a function like:

```python
def get_remaining_editor_parsed_pubmed_links_df(
    editor_parsed_pubmed_links_df: pd.DataFrame,
    existing_editor_parsed_pubmed_links_df: pd.DataFrame
) -> pd.DataFrame:
    ...
```

That function could live in the Python package and would be easy to unit tests. And here in the notebook it could be called.

File: notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb [hunk 5]
```diff
@@ -127,22 +128,138 @@
     "print(\"existing_editor_ids_and_pubmed_url_df length: \", len(existing_editor_ids_and_pubmed_url_df))\n",
     "existing_editor_ids_and_pubmed_url_df.head(3)"
    ],
-   "outputs": [],
+   "outputs": [
```
Comment (de-code) [inline notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb:131]: Oh, how come there are outputs. The pre-commit should have removed it.

See, that is why we perhaps do need the additional check that we had there before.

File: notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb (hunk 2 above)
Comment (HazalCiplak) [reply notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb]: Sure

File: notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb (hunk 5 above)
Comment (de-code) [reply notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb:139]: And it's back again :D

File: data_science_pipeline/utils/bq.py [hunk 6]
```diff
@@ -248,3 +248,66 @@ def run_query_and_save_to_table(  # pylint: disable=too-many-arguments
...
+):
+    client = get_client(project_id=project)
+    view_id = ".".join([project, dataset, view_name])
+    view = bigquery.Table(view_id)
+    view.view_query = query
+    if not does_bq_table_exist(client, project, dataset, view_name):
+        view = client.create_table(view)
+        LOGGER.info("Created %s: %s", view.table_type, str(view.reference))
+    else:
+        client.delete_table(view_id, not_found_ok=True)
```
Comment (de-code) [inline data_science_pipeline/utils/bq.py:284]: I would generally try to avoid deleting a table as part of the general flow (unless it is a cleanup task of course). That is because it is not atomic.

You could look at the
[bigquery-views-manager/](https://github.com/elifesciences/bigquery-views-manager/blob/develop/bigquery_views_manager/update_views.py) that we use for the BigQuery views project. It does indeed use the `CREATE OR REPLACE TABLE` you opted for below.

File: data_science_pipeline/utils/bq.py [hunk 7]
```diff
@@ -248,3 +248,66 @@ def run_query_and_save_to_table(  # pylint: disable=too-many-arguments
...
+        mview_prefix='m'
+):
+    client = get_client(project_id=project)
+    source_view = ".".join([project, dataset, source_view_name])
+    destination_mview_name = "".join([mview_prefix, source_view_name])
+    destination_mview = ".".join([project, dataset, destination_mview_name])
+    LOGGER.info("view %s is been materializing", source_view)
+    job_config = bigquery.QueryJobConfig()
+    sql = '''
+        CREATE OR REPLACE TABLE {destination_mview}
```
Comment (de-code) [inline data_science_pipeline/utils/bq.py:303]: just a comment in case you are not aware that the the BigQuery API has an explicit option to write the result to a destination table (which is used by https://github.com/elifesciences/bigquery-views-manager/blob/develop/bigquery_views_manager/materialize_views.py and the notebooks that do something similar).

Both options are fine as both are atomic.
(More relevant to the BigQuery Views project, the other option provides also access to the result, e.g. how many rows, which is good for logging)

File: data_science_pipeline/utils/bq.py (hunk 6 above)
Comment (de-code) [reply data_science_pipeline/utils/bq.py:284]: Maybe we could even use the `bigquery-views-manager` project to re-use some code?

File: data_science_pipeline/utils/bq.py (hunk 7 above)
Comment (HazalCiplak) [reply data_science_pipeline/utils/bq.py:304]: We have logs already from dataframes so that is OK in that case I guess.

File: data_science_pipeline/utils/bq.py (hunk 6 above)
Comment (HazalCiplak) [reply data_science_pipeline/utils/bq.py:285]: It is actually just deleting the view which is not in use. It is just for seeing the code via bigquery interface without looking at github. Table and materialised view is exist all the time.

Comment (de-code) [reply data_science_pipeline/utils/bq.py:284]: Personally I would always prefer an atomic operation where possible (and re-use code where we can). In this case we already have used a method that is atomic for the BigQuery Views. I don't quite see the advantage of not using that.

But if you really want to stick to this, then I would indicating in the name that it is not atomic in the function name to avoid using for other purposes.

File: data_science_pipeline/utils/bq.py (hunk 7 above)
Comment (de-code) [reply data_science_pipeline/utils/bq.py:303]: The logs are a bit different. The dataframe reflects the data that we add to a table. Whereas the logs when materializing views for the BigQuery views project show information like how many rows the materialized view has. e.g.:

```
[2021-09-20 06:08:20,324] {materialize_views.py:44} INFO - materializing view: prod.v_Person_Latest_Contact_Timestamp -> prod.mv_Person_Latest_Contact_Timestamp
[2021-09-20 06:08:25,785] {materialize_views.py:67} INFO - materialized view: prod.v_Person_Latest_Contact_Timestamp, total rows: 157284, 50662021 bytes processed, took: 5.461s
```

But that kind of logging is more valuable to the materialize views DAG because there one can see the same information for each table, making it relatively easy to identify the views that take to long or are very large.

File: notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb [hunk 8]
```diff
@@ -76,103 +79,226 @@
...
+   "outputs": [],
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "source": [
+    "editors_with_not_currently_updated_info_df = merged_editor_parsed_pubmed_links_df[\n",
+    "    (\n",
+    "        pd.Timestamp.now(tz=pytz.UTC)\n",
```
Comment (de-code) [inline notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb:229]: this probably should be `pd.Timestamp.utcnow()`? Otherwise I guess it will be a local time with the timezone of UTC?

File: notebooks/peerscout/peerscout-create-view-editor-pubmed-ids.ipynb [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    "update_or_create_view(\n",
+    "        client=client,\n",
+    "        view_name=view_name,\n",
+    "        view_query=view_query,\n",
+    "        dataset=dataset\n",
+    ")    "
+   ],
+   "outputs": [
+    {
+     "output_type": "stream",
```
Comment (de-code) [inline notebooks/peerscout/peerscout-create-view-editor-pubmed-ids.ipynb:67]: I'd remove outputs from this notebook too. The nbnotebook strip should do that.
We could look into getting the pre-commit hook for for you.

In any case it would be good to make the linting to check for outputs functional again. Perhaps you could create an issue for it? (It should in general work as long as Jupyter is used via the container. For it to be more flexible the linting can be amended)

**Review by de-code — APPROVED** [review]
